File: v4/rust/iprange-livedb/src/publication/gc_maintenance.rs
//! Explicit discovery and removal of authenticated Windows GC pairs.

use std::path::Path;

use crate::cancellation::CancellationToken;
use crate::error::{Error, Result};
use crate::validation::LocalFileIdentity;

use super::gc;
use super::gc_codec::Payload;
use super::gc_name::{self, Candidate};
use super::maintenance::{
    HousekeepingPayloadIdentity, WindowsHousekeepingCandidateKind, WindowsHousekeepingEntry,
    WindowsHousekeepingList, WindowsHousekeepingRemoval, WindowsHousekeepingSink,
    WindowsHousekeepingSinkControl,
};
use super::namespace::{
    identity_from_local, local_identity as local, Directory, Identity, Name, NamespaceError,
    ScanError, BASENAME_ENCODING_KIND, IDENTITY_KIND,
};
use super::problem::Problem;
use super::{Housekeeping, HousekeepingState};

pub(super) fn list<S: WindowsHousekeepingSink>(
    path: &Path,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> Result<WindowsHousekeepingList> {
    cancellation.check()?;
    let directory = Directory::open(path).map_err(namespace_error)?;
    let directory_identity = local(directory.identity());
    let mut entries = 0u64;
    let scan = directory.scan(|bytes| {
        cancellation.check()?;
        let Some(candidate) = gc_name::candidate(bytes) else {
            return Ok(());
        };
        let entry = inspect(&directory, directory_identity, bytes, candidate);
        deliver(sink, &entry)?;
        entries = entries
            .checked_add(1)
            .ok_or(Error::ArithmeticOverflow("Windows housekeeping entries"))?;
        Ok(())
    });
    match scan {
        Ok(()) => Ok(WindowsHousekeepingList {
            directory_identity,
            entries,
        }),
        Err(ScanError::Namespace(error)) => Err(namespace_error(error)),
        Err(ScanError::Visitor(error)) => Err(error),
    }
}

#[allow(clippy::too_many_arguments)]
pub(super) fn remove(
    path: &Path,
    expected_directory: LocalFileIdentity,
    attempt_id: [u8; 16],
    ordinal: u32,
    expected_envelope: LocalFileIdentity,
    expected_payload: Option<HousekeepingPayloadIdentity>,
    cancellation: &CancellationToken,
) -> Result<WindowsHousekeepingRemoval> {
    cancellation.check()?;
    if attempt_id == [0; 16] {
        return Err(Error::InvalidArgument(
            "housekeeping attempt id must be nonzero",
        ));
    }
    let expected_directory = identity(expected_directory)?;
    let expected_envelope = identity(expected_envelope)?;
    let expected_payload = expected_payload.map(payload).transpose()?;
    let directory = Directory::open(path).map_err(namespace_error)?;
    if directory.identity() != expected_directory {
        return Err(Error::DirectoryIdentityMismatch);
    }
    let envelope_name = gc_name::envelope(attempt_id, ordinal).map_err(namespace_error)?;
    let Some(envelope) =
        gc::open(&directory, envelope_name.clone(), true).map_err(problem_error)?
    else {
        directory.verify().map_err(namespace_error)?;
        directory
            .require_absent(&envelope_name)
            .map_err(cleanup_error)?;
        let inert_name = gc_name::inert(attempt_id, ordinal).map_err(namespace_error)?;
        directory
            .require_absent(&inert_name)
            .map_err(cleanup_error)?;
        return Ok(WindowsHousekeepingRemoval {
            housekeeping: Housekeeping::CrashReappearancePossible,
            visible_housekeeping: Box::new([]),
            cause: None,
        });
    };
    if envelope.identity != expected_envelope {
        return Err(Error::CleanupConflict(
            "GC envelope identity changed before removal",
        ));
    }
    if expected_payload.is_some() && envelope.header.payload != expected_payload {
        return Err(Error::CleanupConflict(
            "GC payload identity changed before removal",
        ));
    }
    cancellation.check()?;
    let retirement = gc::resolve_existing(&directory, envelope);
    Ok(WindowsHousekeepingRemoval {
        housekeeping: retirement.housekeeping,
        visible_housekeeping: retirement.visible.into_iter().collect(),
        cause: retirement.problem,
    })
}

fn inspect(
    directory: &Directory,
    directory_identity: LocalFileIdentity,
    bytes: &[u8],
    candidate: Candidate,
) -> WindowsHousekeepingEntry {
    let basename = encoded_basename(bytes);
    let (candidate_kind, decoded) = match candidate {
        Candidate::Envelope(decoded) => (WindowsHousekeepingCandidateKind::Envelope, decoded),
        Candidate::Inert(decoded) => (WindowsHousekeepingCandidateKind::InertPayload, decoded),
    };
    let identity = Name::new(bytes)
        .ok()
        .and_then(|name| directory.entry(&name).ok().flatten())
        .map(|entry| local(entry.identity));
    let Some((attempt_id, ordinal)) = decoded else {
        return candidate_entry(
            directory_identity,
            candidate_kind,
            &basename,
            identity,
            None,
            None,
            None,
            Some(Problem::cleanup_conflict(
                "Windows housekeeping name is not canonical",
            )),
        );
    };
    let envelope_name = match gc_name::envelope(attempt_id, ordinal) {
        Ok(name) => name,
        Err(error) => {
            return candidate_entry(
                directory_identity,
                candidate_kind,
                &basename,
                identity,
                Some(attempt_id),
                Some(ordinal),
                None,
                Some(Problem::namespace(&error)),
            )
        }
    };
    match gc::open(directory, envelope_name, false) {
        Ok(Some(envelope)) => {
            let observed = gc::observe_pair(directory, &envelope);
            let problem = (observed.state == HousekeepingState::Conflict)
                .then(|| Problem::cleanup_conflict("GC payload names or identities conflict"));
            let artifact = gc::artifact(directory.identity(), &envelope, observed);
            candidate_entry(
                directory_identity,
                candidate_kind,
                &basename,
                identity,
                Some(attempt_id),
                Some(ordinal),
                Some(artifact),
                problem,
            )
        }
        Ok(None) => candidate_entry(
            directory_identity,
            candidate_kind,
            &basename,
            identity,
            Some(attempt_id),
            Some(ordinal),
            None,
            Some(Problem::cleanup_conflict(
                "GC candidate has no authority envelope",
            )),
        ),
        Err(problem) => candidate_entry(
            directory_identity,
            candidate_kind,
            &basename,
            identity,
            Some(attempt_id),
            Some(ordinal),
            None,
            Some(problem),
        ),
    }
}

fn encoded_basename(ascii: &[u8]) -> Box<[u8]> {
    let mut bytes = Vec::with_capacity(ascii.len() * 2);
    for &byte in ascii {
        bytes.extend_from_slice(&u16::from(byte).to_le_bytes());
    }
    bytes.into_boxed_slice()
}

#[allow(clippy::too_many_arguments)]
fn candidate_entry(
    directory_identity: LocalFileIdentity,
    candidate_kind: WindowsHousekeepingCandidateKind,
    basename: &[u8],
    identity: Option<LocalFileIdentity>,
    attempt_id: Option<[u8; 16]>,
    ordinal: Option<u32>,
    artifact: Option<super::HousekeepingArtifact>,
    problem: Option<Problem>,
) -> WindowsHousekeepingEntry {
    WindowsHousekeepingEntry {
        directory_identity,
        candidate_kind,
        basename_encoding: BASENAME_ENCODING_KIND,
        basename: basename.into(),
        identity,
        attempt_id,
        ordinal,
        artifact,
        problem,
    }
}

fn payload(expected: HousekeepingPayloadIdentity) -> Result<Payload> {
    let tuple = expected.tuple;
    if expected.digest.byte_length == 0
        || expected.digest.sha512 == [0; 64]
        || tuple.is_some_and(|tuple| {
            tuple.database_id == [0; 16]
                || tuple.transaction_id == 0
                || tuple.commit_nonce == [0; 16]
        })
    {
        return Err(Error::InvalidArgument(
            "housekeeping payload identity is malformed",
        ));
    }
    Ok(Payload {
        byte_length: expected.digest.byte_length,
        sha512: expected.digest.sha512,
        database_id: tuple.map_or([0; 16], |tuple| tuple.database_id),
        transaction_id: tuple.map_or(0, |tuple| tuple.transaction_id),
        commit_nonce: tuple.map_or([0; 16], |tuple| tuple.commit_nonce),
    })
}

fn deliver<S: WindowsHousekeepingSink>(
    sink: &mut S,
    entry: &WindowsHousekeepingEntry,
) -> Result<()> {
    match sink.entry(entry) {
        Ok(WindowsHousekeepingSinkControl::Continue) => Ok(()),
        Ok(WindowsHousekeepingSinkControl::Stop) => Err(Error::StoppedBySink),
        Err(error) => Err(Error::SinkFailed(Box::new(error))),
    }
}

fn identity(value: LocalFileIdentity) -> Result<Identity> {
    if value.kind != IDENTITY_KIND {
        return Err(Error::InvalidArgument(
            "unsupported Windows housekeeping identity kind",
        ));
    }
    identity_from_local(value).ok_or(Error::InvalidArgument(
        "invalid Windows housekeeping identity",
    ))
}

fn problem_error(problem: Problem) -> Error {
    problem.into_sdk()
}

fn cleanup_error(error: NamespaceError) -> Error {
    match error {
        NamespaceError::Io(source) | NamespaceError::IoAt { source, .. } => Error::Io(source),
        NamespaceError::ForkedHandle => Error::ForkedHandle,
        _ => Error::CleanupConflict("Windows housekeeping ownership changed"),
    }
}

fn namespace_error(error: NamespaceError) -> Error {
    match error {
        NamespaceError::InvalidName => Error::InvalidArgument("invalid Windows housekeeping name"),
        NamespaceError::ForkedHandle => Error::ForkedHandle,
        NamespaceError::Io(source) | NamespaceError::IoAt { source, .. } => Error::Io(source),
        NamespaceError::Unsupported | NamespaceError::CrossFilesystem => {
            Error::Unsupported("housekeeping directory lacks required local NTFS operations")
        }
        _ => Error::CleanupConflict("Windows housekeeping directory entry changed"),
    }
}

File: v4/go/internal/live/gc_maintenance.go
//go:build windows

// Explicit discovery and removal of authenticated Windows GC pairs
// (Rust publication/gc_maintenance.rs). Listing is constant-memory and
// grants no deletion authority: every candidate is inspected against
// its envelope and reported with the exact problem class. Removal
// takes the caller-certified expected identities and resolves the pair
// through the common GC move resolver.

package live

import (
	"github.com/firehol/iprange/v4/go/internal/format"
)

// gcHousekeepingCandidateKind classifies one scanned GC name (Rust
// WindowsHousekeepingCandidateKind).
type gcHousekeepingCandidateKind uint8

const (
	gcCandidateEnvelope gcHousekeepingCandidateKind = iota
	gcCandidateInertPayload
)

// gcHousekeepingEntry is one exact scanned housekeeping candidate
// (Rust WindowsHousekeepingEntry). The basename is the encoded form in
// the platform basename encoding.
type gcHousekeepingEntry struct {
	DirectoryIdentity LocalFileIdentity
	CandidateKind     gcHousekeepingCandidateKind
	BasenameEncoding  uint16
	Basename          []byte
	Identity          *LocalFileIdentity
	AttemptID         *[16]byte
	Ordinal           *uint32
	Artifact          *HousekeepingArtifact
	Problem           *format.Error
}

// gcHousekeepingList is one completed constant-memory housekeeping
// scan (Rust WindowsHousekeepingList).
type gcHousekeepingList struct {
	DirectoryIdentity LocalFileIdentity
	Entries           uint64
}

// gcHousekeepingRemoval is the factual terminal state of one exact
// housekeeping removal (Rust WindowsHousekeepingRemoval).
type gcHousekeepingRemoval struct {
	Housekeeping Housekeeping
	Visible      []HousekeepingArtifact
	Cause        *format.Error
}

// gcHousekeepingPayload is the optional exact content evidence of one
// removal (Rust HousekeepingPayloadIdentity: the digest plus the
// publication tuple).
type gcHousekeepingPayload struct {
	ByteLength    uint64
	SHA512        [64]byte
	DatabaseID    [16]byte
	TransactionID uint64
	CommitNonce   [16]byte
}

// gcListHousekeeping streams one constant-memory housekeeping scan of
// the retained directory (Rust gc_maintenance::list): every GC
// candidate name is inspected against its envelope and delivered to
// the sink.
func gcListHousekeeping(path string, check func() error, sink func(entry *gcHousekeepingEntry) error) (gcHousekeepingList, error) {
	if err := check(); err != nil {
		return gcHousekeepingList{}, err
	}
	dir, err := OpenDirectory(path)
	if err != nil {
		return gcHousekeepingList{}, gcNamespaceProblem(err)
	}
	defer dir.Close()
	directoryIdentity := LocalFileIdentityFromDeviceInode(dir.Identity().device, dir.Identity().inode)
	var entries uint64
	err = dir.Scan(func(bytes []byte) error {
		if err := check(); err != nil {
			return err
		}
		candidate := gcCandidateOf(bytes)
		if !candidate.matched {
			return nil
		}
		entry := gcInspectHousekeeping(dir, directoryIdentity, bytes, candidate)
		if err := sink(entry); err != nil {
			return err
		}
		entries++
		return nil
	})
	if err != nil {
		return gcHousekeepingList{}, err
	}
	return gcHousekeepingList{DirectoryIdentity: directoryIdentity, Entries: entries}, nil
}

// gcInspectHousekeeping inspects one GC candidate (Rust
// gc_maintenance::inspect): the canonical decode, the envelope open,
// the observed pair, and the artifact projection.
func gcInspectHousekeeping(dir *Directory, directoryIdentity LocalFileIdentity, bytes []byte, candidate gcCandidate) *gcHousekeepingEntry {
	encoded := gcEncodedBasename(bytes)
	kind := gcCandidateEnvelope
	if !candidate.envelope {
		kind = gcCandidateInertPayload
	}
	identity := gcEntryIdentity(dir, bytes)
	attempt, ordinal, decoded := candidate.attempt, candidate.ordinal, candidate.decoded
	if !decoded {
		return gcHousekeepingCandidateEntry(directoryIdentity, kind, encoded, identity, nil, nil, nil,
			gcCleanupConflict("Windows housekeeping name is not canonical"))
	}
	envelopeName, err := gcEnvelopeName(attempt, ordinal)
	if err != nil {
		return gcHousekeepingCandidateEntry(directoryIdentity, kind, encoded, identity, &attempt, &ordinal, nil,
			gcNamespaceProblem(err))
	}
	envelope, openErr := gcOpen(dir, envelopeName, false)
	if openErr != nil {
		return gcHousekeepingCandidateEntry(directoryIdentity, kind, encoded, identity, &attempt, &ordinal, nil, gcProblemOf(openErr))
	}
	if envelope == nil {
		return gcHousekeepingCandidateEntry(directoryIdentity, kind, encoded, identity, &attempt, &ordinal, nil,
			gcCleanupConflict("GC candidate has no authority envelope"))
	}
	observed := gcObservePair(dir, envelope)
	var problem *format.Error
	if observed.state == HousekeepingConflict {
		problem = gcCleanupConflict("GC payload names or identities conflict")
	}
	artifact := gcArtifact(dir.Identity(), envelope, observed)
	return gcHousekeepingCandidateEntry(directoryIdentity, kind, encoded, identity, &attempt, &ordinal, &artifact, problem)
}

// gcEncodedBasename converts one scanned ASCII name to the encoded
// basename of the platform (Rust gc_maintenance::encoded_basename:
// ASCII to UTF-16LE on windows).
func gcEncodedBasename(ascii []byte) []byte {
	return gcNameBytes(string(ascii))
}

// gcEntryIdentity looks up one scanned name (Rust directory.entry).
func gcEntryIdentity(dir *Directory, bytes []byte) *LocalFileIdentity {
	entry, present, err := dir.Entry(string(bytes))
	if err != nil || !present {
		return nil
	}
	value := LocalFileIdentityFromDeviceInode(entry.Identity.device, entry.Identity.inode)
	return &value
}

// gcHousekeepingCandidateEntry builds one entry with the optional
// decodes and problem (Rust gc_maintenance::candidate_entry).
func gcHousekeepingCandidateEntry(directoryIdentity LocalFileIdentity, kind gcHousekeepingCandidateKind, basename []byte, identity *LocalFileIdentity, attempt *[16]byte, ordinal *uint32, artifact *HousekeepingArtifact, problem *format.Error) *gcHousekeepingEntry {
	return &gcHousekeepingEntry{
		DirectoryIdentity: directoryIdentity,
		CandidateKind:     kind,
		BasenameEncoding:  uint16(gcBasenameEncodingValue()),
		Basename:          basename,
		Identity:          identity,
		AttemptID:         attempt,
		Ordinal:           ordinal,
		Artifact:          artifact,
		Problem:           problem,
	}
}

// gcRemoveHousekeeping resolves and best-effort removes one exact GC
// pair (Rust gc_maintenance::remove): the directory identity must
// match, the envelope must exist with the expected identity and
// payload, and the common resolver runs the move.
func gcRemoveHousekeeping(path string, expectedDirectory LocalFileIdentity, attemptID [16]byte, ordinal uint32, expectedEnvelope LocalFileIdentity, expectedPayload *gcHousekeepingPayload, check func() error) (gcHousekeepingRemoval, error) {
	if err := check(); err != nil {
		return gcHousekeepingRemoval{}, err
	}
	if attemptID == [16]byte{} {
		return gcHousekeepingRemoval{}, &format.Error{Code: format.CodeInvalidArgument, Detail: "housekeeping attempt id must be nonzero"}
	}
	dir, err := OpenDirectory(path)
	if err != nil {
		return gcHousekeepingRemoval{}, gcNamespaceProblem(err)
	}
	defer dir.Close()
	if LocalFileIdentityFromDeviceInode(dir.Identity().device, dir.Identity().inode) != expectedDirectory {
		return gcHousekeepingRemoval{}, &format.Error{Code: format.CodeConflict, Detail: "housekeeping directory identity mismatch"}
	}
	envelopeName, err := gcEnvelopeName(attemptID, ordinal)
	if err != nil {
		return gcHousekeepingRemoval{}, gcNamespaceProblem(err)
	}
	envelope, err := gcOpen(dir, envelopeName, true)
	if err != nil {
		return gcHousekeepingRemoval{}, gcCleanupConflict("GC envelope is not resolvable")
	}
	if envelope == nil {
		if err := dir.Verify(); err != nil {
			return gcHousekeepingRemoval{}, gcNamespaceProblem(err)
		}
		if err := dir.RequireAbsent(envelopeName); err != nil {
			return gcHousekeepingRemoval{}, gcCleanupConflict("Windows housekeeping ownership changed")
		}
		inertName, err := gcInertName(attemptID, ordinal)
		if err != nil {
			return gcHousekeepingRemoval{}, gcNamespaceProblem(err)
		}
		if err := dir.RequireAbsent(inertName); err != nil {
			return gcHousekeepingRemoval{}, gcCleanupConflict("Windows housekeeping ownership changed")
		}
		return gcHousekeepingRemoval{Housekeeping: HousekeepingCrashReappearancePossible, Cause: nil}, nil
	}
	if LocalFileIdentityFromDeviceInode(envelope.identity.device, envelope.identity.inode) != expectedEnvelope {
		return gcHousekeepingRemoval{}, gcCleanupConflict("GC envelope identity changed before removal")
	}
	if expectedPayload != nil && !gcPayloadMatchesExpected(envelope.header.payload, expectedPayload) {
		return gcHousekeepingRemoval{}, gcCleanupConflict("GC payload identity changed before removal")
	}
	if err := check(); err != nil {
		return gcHousekeepingRemoval{}, err
	}
	retired := gcResolveExisting(dir, envelope)
	return gcHousekeepingRemoval{
		Housekeeping: retired.housekeeping,
		Visible:      gcSliceOf(retired.visible),
		Cause:        gcProblemOrNil(retired.problem),
	}, nil
}

// gcPayloadMatchesExpected compares one envelope payload against the
// caller-certified expectation (Rust equality of Option<Payload>).
func gcPayloadMatchesExpected(payload *gcPayload, expected *gcHousekeepingPayload) bool {
	if payload == nil {
		return false
	}
	if payload.byteLength != expected.ByteLength || payload.sha512 != expected.SHA512 {
		return false
	}
	if payload.databaseID == [16]byte{} {
		return expected.DatabaseID == [16]byte{} && expected.TransactionID == 0 && expected.CommitNonce == [16]byte{}
	}
	return payload.databaseID == expected.DatabaseID &&
		payload.transactionID == expected.TransactionID &&
		payload.commitNonce == expected.CommitNonce
}

// gcSliceOf folds one optional artifact into the removal ledger (Rust
// visible.into_iter().collect()).
func gcSliceOf(artifact *HousekeepingArtifact) []HousekeepingArtifact {
	if artifact == nil {
		return nil
	}
	return []HousekeepingArtifact{*artifact}
}
